### ISSUE TITLE
Adding member via NFT

### ISSUE THREAD

**Issue by c4sp3r-sol** [cardinal-labs/hydra-ui#13]
In the Metaplex Docs, its notes that the NFT method is another way to add members to the wallet, is this currently wokring/being worked on?

"With this model membership is tied to an NFT mint address instead of static public address. Each NFT mint address can still have a different quantity of shares as in the Wallet model. The greatest benefit of this model is it effectively enables the simple transfer of rights to future distributions to any wallet owner that holds the given NFT."

**Comment by Swertin** [comment]
Looks like the code for this is pretty similar to the current method, just implementing a few inputs for selecting and using the selected membership model [here](https://github.com/cardinal-labs/hydra-ui/blob/2a0ae869e7bd4e2ec8b74954ee79ad9bdf8f0c69/pages/create/index.tsx#L78), and adding a check for the distribution method/ix.

```
const nftMintPublicKey = "nftMintPublicKey";

const init = await fanoutSdk.initializeFanout({
    totalShares: 100,
    name: `Test${Date.now()}`,
    membershipModel: MembershipModel.NFT,
});

const {membershipAccount} = await fanoutSdk.addMemberNft({
    fanout: init.fanout,
    fanoutNativeAccount: init.nativeAccount,
    membershipKey: nftMintPublicKey,
    shares: 10
});

// Add members until sum of shares = totalShares
...
```

One challenge is that the assumption would be you're intending to setup a fanout for an entire nft collection, being potentially 10,000 nfts (ignoring compression support). So with the current operational model of this UI, it would need to get upgraded to split the initialization flow and adding members flow, because we definitely can't fit it all into 1 transaction. 

Honestly, I feel like the best solution would be to split the initialization from adding members on create.tsx and moving member configuration to the initialized fanout page. Knowing total shares is necessary at the initialization step, so that's unchanged. The only place that a wallet cares if all members have been added (sum of member shares = total shares) is when a distribution event is attempted.

This way, we can batch adding members with simpler logic than having to account for the initialize fanout ix, which solves issue #4. It also provides for a better UX when it comes to membership management, being that they could add or remove members in a place that they've already done it before (after initializing and redirecting to their initialized page).

This should also have some way to import csv or a csv formatted membership list (or json, but json is more finnicky when creating for non-developers), and some display for "total shares | sum of current membership shares" in order to help the user understand if they're missing some members if distributions fail.

Overall, the larger challenge I see will definitely be token based, which should also be considered as a feature addition but probably on a separate issue.

I do see there being some potentially wasted time for handling bulk nft adds (1000+) if collection mints are ever supported for hydra, but I do think this supporting work would improve UX for wallet adds as well and should be worthwhile to implement.

Open to any other suggestions for how to implement before getting to work.

**Comment by jpbogle** [comment]
Agreed with @Swertin that the main use-case is probably NFT collections which is a large number of transactions

this is an example of handling large number (thousands) of transactions in a UI, and as @Swertin mentioned one way to do it would allows the user to pass in a csv of members for example
https://github.com/cardinal-labs/cardinal-staking-ui/blob/main/components/admin/TransactionExecutor.tsx

That then begs the question, where is the user going to get their CSV of members if they are not a dev. We have an index of all NFTs but doing that in the UI also would introduce a dependency on off-chain infra which we do not currently want.

Curious @c4sp3r-sol if you have a main use-case you had in mind for token based membership

**Comment by gchatz22** [comment]
Agreed with @jpbogle on the `TransactionExecutor` example he linked above. We've had dealt with similar issues in our UIs where for example stake pool owners want to set different stake multipliers for each mint in the collection through the admin UI portal.

> That then begs the question, where is the user going to get their CSV of members if they are not a dev. We have an index of all NFTs but doing that in the UI also would introduce a dependency on off-chain infra which we do not currently want.

also agree here, the way I can imagine this would be if the membership is tied to an NFT mint address as @c4sp3r-sol mentioned (which is currently not implemented fwiw). A question then is how do you store all these mint addresses now because you cant just put them in a `Vec<Pubkey>` right? I believe then a model like the Candy Machine config lines should be followed where mint addresses are being added as raw account data in the account.

**Comment by c4sp3r-sol** [comment]
@jpbogle the main use case I was looking at is for DAO centered NFT collections. Having a wallet that takes royalty and fans it out to all active holders by tying it with the NFT address itself. This ensures whoever buys the nft off secondary will receive their share in DAO efforts. Understanding that there are limits to the amount of transaction, I see how this may be a more difficult task on the UI then previously thought.

In the past, I have manually created fanout using a chrome wallet through csv importing so I do think having this being a method would greatly reduce time making the task at hand much easier; just a matter of it this is possible with the current build.

**Comment by c4sp3r-sol** [comment]
@jpbogle the main use case I had in mind was the ability to track holders based on NFT ID so distribution of coins could be streamlined to holders without updating any wallet parameters. 

Given the constraints on transaction limit, I see how this could cause issue with collection sizes being too large. In the past I have manually performed fanouts through coin98 wallet where I would build out a CSV to copy and past as distribution which is tedious to perform.

I have not looked far into the repo and do not know the specific constraints, but find that importing a CSV containing all nft address and shares would work for the initialization of the wallet; with each NFT holding the same share value.

**Comment by Swertin** [comment]
Imo, the most common use case is as @c4sp3r-sol mentions, which is equal shares where one nft equals one share out of total nft collection size shares (along with token allocation distribution for ICO/IDO/investors). That being said, by importing as a CSV, the user isn't limited in that they can assign more shares to specific mint addresses if they want to. Batch data entry on excel is a much lower barrier to entry than any other current solutions, and it would be very simple to get a collection hashlist from Metaboss or other external tools, paste into excel, and paste a 1 in each row next to the corresponding mint address. 

Obviously it would be great to build a one stop shop for users to go from collection address to functioning fanout. That being said, that gets built by starting with just the functionality to either import an externally created CSV (which would also work for wallet membership models) after the fanout is initialized or go row by row through modifying and extending the current interface
